mava/wrappers/gigastep.py

```diff
@@ -0,0 +1,301 @@
+
+        action = jnp.concatenate([action, state.adversary_action], axis=0, dtype=jnp.int16)
+
+        obs, env_state, rewards, dones, ep_done = self._env.step(state.state, action, step_key)
+
+        # cut out the rewards,dones of the adversary
+        rewards, dones = (
+            rewards[: self.num_agents],
+            dones[: self.num_agents],
+        )
+
+        obs_team1, state_team1, obs_team2, state_team2 = self._split_obs_and_state(obs, env_state)
+
+        # take the actions of the adversary and cache it before returning the new state
+        adversary_actions = self.adversary_policy(obs_team2, state_team2, adversary_key)
+
+        obs = self._create_observation(obs_team1, adversary_actions, obs, state)
+
+        step_type = jax.lax.select(ep_done, StepType.LAST, StepType.MID)
+
+        current_winner = ep_done & self.won_episode(env_state)
+
+        ts = TimeStep(
+            step_type=step_type,
+            reward=rewards,
+            discount=1.0 - dones,
+            observation=obs,
+            extras={"won_episode": current_winner},
+        )
+        return GigastepState(env_state, key, state.step + 1, adversary_actions), ts
+
+    def _create_observation(
+        self,
+        obs: Array,
+        adversary_actions: Array,
+        obs_full: Array,
+        state: GigastepState,
+    ) -> Union[Observation, ObservationGlobalState]:
+        """Create an observation from the raw observation and environment state."""
+        obs_data = {
+            "agents_view": obs,
+            "action_mask": self.action_mask(),
+            "step_count": jnp.repeat(state.step, self.num_agents),
+        }
+
+        if self.has_global_state:
+            obs_data["global_state"] = self.get_global_state(obs_full, adversary_actions)
+            return ObservationGlobalState(**obs_data)
+        else:
+            return Observation(**obs_data)
+
+    def action_mask(self) -> Array:
+        """Get action mask for each agent."""
+        return jnp.ones((self.num_agents, self._env.n_actions))  # all actions are valid
+
+    def get_global_state(self, obs: Array, adversary_actions: Array) -> Array:
+        """
+        Combines observations from all agents,
+        adversary actions, and adversary observations
+        to create a global state for the environment.
+
+        Args:
+            obs (Array): The observations of all agents.
+            adversary_actions (Array): The actions for the adversary.
+
+        Returns:
+            global_obs (Array): The global observation.
+        """
+        # todo: still unsure if adversary actions and observations is too much information in
+        # the global Needs to be tested once we have better heuristics for adversaries.
+        global_obs = jnp.concatenate(obs, axis=0)
+        # add the adversary actions to the end of the global observation
+        global_obs = jnp.concatenate([global_obs, adversary_actions], axis=0)
```

Comment:
the get_global_obs() provided by the env is an RGB image so i dont think using it is a good idea. i think ruan suggested to go with concatenating the vectors untill cnns become supported (they are?)
and a few comments up @sash-a suggested either always adding (or removing) the actions in order to make our results easily reproducible

Reply:
I see I thought it would give same dimension as for the agent's view, then yeah concatenation is the best way for now.
@sash-a please check my above comment (correct me if I'm wrong), I guess if we want to set supporting one way to get the global_obs I would vote for just using the `GlobalWrapper` since the addition of the `last actions` should be optional and that's not the way we do it usually, what do you think?🤔